File: app/src/main/java/wmsdf/cl/exp4/getwallpaper/util/WallpaperUtils.java
package wmsdf.cl.exp4.getwallpaper.util;

import android.Manifest;
import android.app.WallpaperManager;
import android.content.Context;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.ParcelFileDescriptor;

import androidx.annotation.RequiresApi;
import androidx.core.app.ActivityCompat;

import java.io.FileDescriptor;

public class WallpaperUtils {

    public static Drawable getWallpaperDrawable(Context context) {
        WallpaperManager wpm = WallpaperManager.getInstance(context);
        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            return null;
        }
        return wpm.getDrawable();
    }

    @RequiresApi(24)
    public static Drawable getLockWallpaperDrawable(Context context){
        WallpaperManager wpm = WallpaperManager.getInstance(context);
        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            return null;
        }
        ParcelFileDescriptor mParcelFileDescriptor = wpm.getWallpaperFile(WallpaperManager.FLAG_LOCK); // 获取锁屏壁纸
        if(mParcelFileDescriptor == null){
            return getWallpaperDrawable(context);
        }
        FileDescriptor fileDescriptor = mParcelFileDescriptor.getFileDescriptor();
        Bitmap image = BitmapFactory.decodeFileDescriptor(fileDescriptor);//获取Bitmap类型返回值
        try {
            mParcelFileDescriptor.close();
        } catch(Exception e) {
            android.util.Log.e("app/WallpaperUtils","mParcelFileDescriptor.close() error");
        }
        return new BitmapDrawable(context.getResources(), image);
    }

}

File: kotlin/src/main/java/wmsdf/cl/exp4/getwallpaper/kotlin/util/WallpaperUtils.kt
package wmsdf.cl.exp4.getwallpaper.kotlin.util

import android.Manifest
import android.annotation.SuppressLint
import android.app.WallpaperManager
import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.os.ParcelFileDescriptor
import android.util.Log
import androidx.annotation.RequiresApi
import androidx.core.app.ActivityCompat
import java.io.FileDescriptor

object WallpaperUtils {

    private fun hasWallpaperAccess(context: Context): Boolean {
        return PermissionUtils.checkPermission(context, Manifest.permission.READ_EXTERNAL_STORAGE);
    }

    @SuppressLint("MissingPermission")
    fun getWallpaperDrawable(context: Context): Drawable? {
        val wpm = WallpaperManager.getInstance(context)
        if (!hasWallpaperAccess(context)) {
            return null
        }
        return wpm.drawable
    }

    @RequiresApi(24)
    @SuppressLint("MissingPermission")
    fun getLockWallpaperDrawable(context: Context): Drawable? {
        val wpm = WallpaperManager.getInstance(context)
        if(!hasWallpaperAccess(context)) {
            return null
        }
        val parcelFileDescriptor: ParcelFileDescriptor?
            = wpm.getWallpaperFile(WallpaperManager.FLAG_LOCK)
        if(parcelFileDescriptor == null) {
            return getWallpaperDrawable(context)
        }
        val fileDescriptor = parcelFileDescriptor.fileDescriptor;
        val image = BitmapFactory.decodeFileDescriptor(fileDescriptor)
        try {
            parcelFileDescriptor.close()
        } catch(e: Exception) {
            Log.e("app/WallpaperUtils", "mParcelFileDescriptor.close() error")
        }
        return BitmapDrawable(context.resources, image)
    }

}